sycl/source/detail/graph_impl.hpp

```diff
@@ -43,6 +43,9 @@ class node_impl {
   /// Command group object which stores all args etc needed to enqueue the node
   std::unique_ptr<sycl::detail::CG> MCommandGroup;
 
+  /// Protects all the fields that can be changed by class' methods.
+  mutable std::recursive_mutex MMutexNode;
```

Comment:
I'm a bit concerned we're overcomplicating things here with mutexes on the nodes themselves. It is a problem that the nodes are shared between modifiable and executable graphs but I think it's preferable to solve that problem rather than use mutexes here. IMO, nodes should be modified only by the owning graph, so we need to be copying them into the executable graph, or some other solution that separates them. This should be fairly straightforward apart from rebuilding the successors/predecessors, but that's not an unsolvable problem.

However, adding that to this task would increase the scope too much. As a stop-gap could we just always lock the modifiable graph when we are dealing with shared data (nodes)? Using a `std::shared_mutex` for aquiring read/write locks.

For recursive functions same as my other comment we should be having private impls and just lock and call them from the public function. Any other private functions that call them should be then using the private impl to avoid the need for recursive mutexes.